CRUD Volunteer model

tirthajyoti-ghosh opened the pull request:
- Created a `volunteer` model.
- Created `CRUD` without `#index` as requested.
- Reformatted directory structure in `app/controllers` directory for API versioning. Changed `routes.rb` file accordingly.
- Added most basic validation in `volunteer` model.
- Used `faker` gem to populate DB with data.
- Changed `db/seeds.rb` to run faker

tirthajyoti-ghosh committed:
added volunteer model

created image_url, name and job_desc for volunteer model. reformatted directory structure to accomodate api versioning.

tirthajyoti-ghosh committed:
added #update and #destroy

tirthajyoti-ghosh committed:
removed #index

tirthajyoti-ghosh committed:
enabled get route

nikasulo commented on config/routes.rb:
```diff
@@ -1,4 +1,9 @@
 Rails.application.routes.draw do
+    namespace 'api' do
+        namespace 'v1' do
+            resources :volunteers, except: :index
+        end
+    end
```
This is very thoughful! @tirthajyoti-ghosh great job :rocket:

tirthajyoti-ghosh replied:
Thanks.

nikasulo commented on app/models/volunteer.rb:
```diff
@@ -0,0 +1,3 @@
+class Volunteer < ApplicationRecord
+  validates :image_url, :name, :job_desc, presence: true  
```
Perfect!

tirthajyoti-ghosh replied:
Thanks.

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,65 @@
+      private
+
+      def volunteer_params
+        params.require(:volunteer).permit(:image_url, :name, :job_desc)
```
Great use of strong parameters here!

tirthajyoti-ghosh replied:
Thanks.

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,65 @@
+      def get
+        volunteers_json = fetchVolunteerJson;
+
+        respond_to do |format|
+            format.json do
+                render(json: volunteers_json)
+            end
+        end
```
There is a custom json response module located in `/app/controllers/concerns` which is a helper for rendering son. so in this case respond_to and the render method are redundant. This can simply read `json_response(volunteers_json)`

tirthajyoti-ghosh replied:
Done

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,65 @@
+  module V1
+    class VolunteersController < ApplicationController
+      def show
+        volunteer = Volunteer.find(params[:id])
```
This can be abstracted into a `find_volunteer` private method and reused in the update and delete actions. And you can add a before filter action, `before_action :find_user, only: %i[show, delete, update]` to run the method before those actions

tirthajyoti-ghosh replied:
Done.

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,65 @@
+  module V1
+    class VolunteersController < ApplicationController
+      def show
+        volunteer = Volunteer.find(params[:id])
```
Also, use the methods with the bang operator(!) such as `.find!`, use `.create!(object_params)` instead of `.new` and `save` and so on. The reason for this is I wrote a custom error handling module that returns json errors. You can find it in `app/controllers/concerns`, it rescues from all exceptions but methods without the bang operator do not trigger exceptions and we need them to so the appropriate JSON messages are returned to the client

tirthajyoti-ghosh replied:
Done.

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,65 @@
+        volunteer = Volunteer.new(volunteer_params)
+
+        if volunteer.save
+          render json: { status: 'SUCCESS', message: 'Created volunteer', data: volunteer }, status: :ok
+        else
+          render json: { status: 'ERROR', message: 'Volunteer not saved', data: volunteer.errors }, status: :unprocessable_entity
+        end
+      end
+
+      def update
+        volunteer = Volunteer.find(params[:id])
+
+        if volunteer.update(volunteer_params)
+          render json: { status: 'SUCCESS', message: 'Updated volunteer', data: volunteer }, status: :ok
+        else
+          render json: { status: 'ERROR', message: 'Volunteer not updated', data: volunteer.errors }, status: :unprocessable_entity
+        end
```
Replace all render methods with the custom response module described below. Also use bang operators as described above, this will eliminate your need for conditionals

tirthajyoti-ghosh replied:
Done.

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,65 @@
+        volunteer = Volunteer.find(params[:id])
+        volunteer.destroy
+
+        render json: { status: 'SUCCESS', message: 'Deleted volunteer', data: volunteer }, status: :ok
```
Delete actions should render no content. Set the `head: :no_content`. The render method isn't necessary here as no json is rendered/expected. Move the find statement into a private method as described above.

tirthajyoti-ghosh replied:
Done

nikasulo reviewed (changes requested):
# Great job! @tirthajyoti-ghosh :rocket: 
Really great effort and thanks for the changes to the routes file. Please take a gander at the comments I've written below, make the appropriate changes and call attention to it when that's done.

##### Think I missed something? Need some help with this or any other projects? Be sure to reach out to me on [Twitter](https://twitter.com/oluwadamilareo_), [Email](mailto:open.source@oluwadamilareolusakin), or [Instagram](https://instagram.com/oluwadamilare_olusakin), and I will respond to you as soon as I can.

tirthajyoti-ghosh committed:
added exception_handler and response files in concerns

These files are not created by me. I simply added them from the master branch manually into volunteer branch.

tirthajyoti-ghosh committed:
removed render and ifs, abstracted #show

made use of controller concerns for json response and error handling, made the functionality of #show abstract

tirthajyoti-ghosh commented:
Changes after code review:
- Abstracted `#show`, especially the finding volunteers functionality.
- Removed `render` to take the help of the `json_response` in  `controllers/concerns`
- Removed conditionals and modified methods with their bang (`!`) version to take advantage of `exception handler` in `controllers/concerns`

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,57 @@
+      def get
+        volunteers_json = fetchVolunteerJson;
+
+        respond_to do |format|
+          format.json do
+            render(json: volunteers_json)
+          end
+        end
```
You left out the json_response method.

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,57 @@
+      private
+
+      def find_volunteer
+        Volunteer.find(params[:id])
```
Use `.find!`

nikasulo replied:
This should read @volunteer = Volunteer.find!(params[:id]) so the variable can be reused in update and delete functions

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,57 @@
+      end
+
+      def destroy
+        find_volunteer.destroy!
```
You don't need `.destroy!` destroy is fine. `.find!` already raises an error if the object cannot be found

nikasulo replied:
@volunteer.destroy will be fine

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,57 @@
+module Api
+  module V1
+    class VolunteersController < ApplicationController
+      include Response
```
You don't need to include response and exception handler. They are already inherited from the application controller as a result of the line above

nikasulo replied:
The find_volunteers method should be added as a before filter method here instead of being used in the update and destroy methods. It's best practice to have `before_action :find_volunteer, only: %![update destroy]`

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,57 @@
+      end
+
+      def create
+        json_response(Volunteer.create!(volunteer_params))
```
create a variable and before you render it as json. Try not to render the create function. It works but it erases consistency and readability

nikasulo commented on app/controllers/api/v1/volunteers_controller.rb:
```diff
@@ -0,0 +1,57 @@
+      end
+
+      def update
+        json_response(find_volunteer.update!(volunteer_params))
```
The update function should also return no content in the head

nikasulo replied:
@volunteer.update(volunteer_params) is sufficient, the @volunteer variable will be set before this action runs by the code you are to place at the top of the file

nikasulo reviewed (changes requested):
# Nice refactor!
Please consider my comments below

tirthajyoti-ghosh committed:
Merge branch 'development' of https://github.com/schoolclosures/covid-19-school-closures-server into volunteer

tirthajyoti-ghosh committed:
added before_action, modified methods accordingly

tirthajyoti-ghosh committed:
enabled #index, changed .find to .find_by!

tirthajyoti-ghosh committed:
refactored serializers

nikasulo reviewed (approved):
# Really great work! @tirthajyoti-ghosh 
Thank you for all the help, you have done an awesome job so far and you deserve applause.

##### Think I missed something? Need some help with this or any other projects? Be sure to reach out to me on [Twitter](https://twitter.com/oluwadamilareo_), [Email](mailto:open.source@oluwadamilareolusakin), or [Instagram](https://instagram.com/oluwadamilare_olusakin), and I will respond to you as soon as I can.
